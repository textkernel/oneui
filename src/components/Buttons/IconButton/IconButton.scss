@use '../Button/Button';

.IconButton {
    &--size {
        &_small {
            padding: var(--space-75);
        }
        &_medium {
            padding: var(--space-100);
        }
        &_large {
            padding: var(--space-125);
        }
    }
}

.IconButton__icon {
    &--size {
        &_small {
            width: var(--space-150);
            height: var(--space-150);
        }
        &_medium {
            width: var(--space-200);
            height: var(--space-200);
        }
        &_large {
            width: var(--space-250);
            height: var(--space-250);
        }
    }
}
