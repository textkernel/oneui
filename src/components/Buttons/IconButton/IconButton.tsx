import * as React from 'react';
import { bem } from '../../../utils';
import { ButtonContext, ButtonType, ButtonVariant, Size } from '../../../constants';
import { Button } from '../Button';
import styles from './IconButton.scss';
import { LoadingSpinner } from '../../LoadingSpinner';

export interface Props extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {
    /** The icon of the Button */
    children: React.ReactElement;
    /** The size of the button */
    size?: Size;
    /** Define the button variant, eg. filled, outlined  */
    variant?: ButtonVariant;
    /** Define the button context, eg. primary, secondary, critical */
    context?: ButtonContext;
    /** Whether or not to show the button in loading state */
    isLoading?: boolean;
    /** Providing a href will render an <a> element, styled as a button. */
    href?: string;
    /** Type of the button */
    type?: ButtonType;
}

const { block, elem } = bem('IconButton', styles);

export const IconButton = React.forwardRef<HTMLInputElement, Props>(
    (
        {
            children,
            disabled = false,
            size = 'medium',
            variant = 'filled',
            context = 'secondary',
            isLoading = false,
            type = 'button',
            href,
            ...rest
        },
        ref
    ) => {
        const icon = React.cloneElement(
            isLoading ? (
                <LoadingSpinner context={variant === 'filled' ? 'inverted' : context} />
            ) : (
                children
            ),
            {
                size,
                ...elem('icon', {
                    size,
                }),
            }
        );

        return (
            <Button
                ref={ref}
                disabled={disabled}
                size={size}
                variant={variant}
                context={context}
                href={href}
                type={type}
                {...rest}
                {...block({ size })}
                isLoading={false}
            >
                {icon}
            </Button>
        );
    }
);

IconButton.displayName = 'IconButton';
