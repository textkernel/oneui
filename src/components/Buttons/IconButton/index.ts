export { IconButton } from './IconButton';
export { Props as IconButtonProps } from './IconButton';
