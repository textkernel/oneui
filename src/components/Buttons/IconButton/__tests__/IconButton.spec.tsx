import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { FaPlus } from 'react-icons/fa';
import { IconButton } from '../IconButton';

describe('IconButton', () => {
    it('should render icon button correctly', () => {
        const view = render(
            <IconButton>
                <FaPlus aria-label="plus" />
            </IconButton>
        );

        expect(view.container).toMatchSnapshot();
        expect(screen.getByLabelText('plus')).toBeInTheDocument();
    });

    it('should add classes when props are changed', () => {
        const view = render(
            <IconButton size="large">
                <FaPlus aria-label="plus" />
            </IconButton>
        );

        const button = screen.getByRole('button');
        expect(button).toHaveClass('IconButton--size_large');
        expect(view.container).toMatchSnapshot();
    });

    it('should call click callback correctly', async () => {
        const onClickMock = jest.fn();
        const user = userEvent.setup();

        render(
            <IconButton onClick={onClickMock}>
                <FaPlus aria-label="plus" />
            </IconButton>
        );

        await user.click(screen.getByLabelText('plus'));

        expect(onClickMock).toHaveBeenCalled();
    });
});
