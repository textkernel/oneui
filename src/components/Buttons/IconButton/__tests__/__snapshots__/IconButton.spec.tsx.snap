// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IconButton should add classes when props are changed 1`] = `
<div>
  <button
    class="Button Button--size_large Button--variant_filled Button--context_secondary IconButton--size_large"
    style="width: auto;"
    type="button"
  >
    <svg
      aria-label="plus"
      class="IconButton__icon--size_large"
      fill="currentColor"
      height="large"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 448 512"
      width="large"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      />
    </svg>
  </button>
</div>
`;

exports[`IconButton should render icon button correctly 1`] = `
<div>
  <button
    class="Button Button--variant_filled Button--context_secondary IconButton--size_medium"
    style="width: auto;"
    type="button"
  >
    <svg
      aria-label="plus"
      class="IconButton__icon--size_medium"
      fill="currentColor"
      height="medium"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 448 512"
      width="medium"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      />
    </svg>
  </button>
</div>
`;
